File: sdk/web/src/animation.ts
// The pager's animation, read from the generated artifact.
//
// Nothing here spells out a speed or a curve constant, for the same reason
// nothing in `theme.ts` spells out a colour: `ANIMATION` is serialized from the
// Rust the pager itself animates on
// (`xai-grok-pager-render/src/theme/animation.rs`), so a tweak there fails
// `theme::tokens::tests::generated_theme_tokens_match_the_pager_palette` until
// it is regenerated, and lands here when it is.
//
// What the artifact deliberately does not carry is `wave_rows`. The pager
// paints a running rail as a *column of terminal cells* and offsets each cell's
// phase by its row; `wave_rows` is how many rows one full wave spans. A rail
// here is a single element, so it is the pager's row 0 and there is no row
// index to offset. `wave_rows` is therefore terminal geometry rather than a
// preference this client quietly fails to honour, which is also why it stays a
// `pager.toml` knob — `SettingSurface::Terminal` defines itself as exactly
// that, "a column width, the frame cadence".
import { createSignal, onCleanup, onMount } from "solid-js";
import { ANIMATION } from "@grok-build/theme";

/**
 * Frame cadence. Every speed below is radians *per tick*, so the cadence is
 * part of the timing.
 *
 * This is the pager's shipped default. `pager.toml` can retune the terminal's
 * own cadence, and nothing carries that to a browser — which the generated
 * artifact says in as many words.
 */
export const TICKS_PER_SECOND = ANIMATION.ticks_per_second;

/**
 * `sin²(tick·speed)` — the pager's `pulse_brightness`.
 *
 * Everything driven by one tick pulses in unison, which is what makes a row of
 * bullets read as one animation rather than several.
 */
export function pulseBrightness(tick: number, speed: number): number {
  const s = Math.sin(tick * speed);
  return s * s;
}

/**
 * The running accent wave: tool rails, running bullets, the running verb-group
 * diamond.
 *
 * The pager's `wave_brightness` at row 0. A rail here is one element, so there
 * is no per-row phase offset to apply — see the note at the top of this file.
 */
export function waveBrightness(tick: number): number {
  return pulseBrightness(tick, ANIMATION.wave_speed);
}

/**
 * The "waiting on you" pulse: the pager's `waiting_brightness`, a pulse lifted
 * onto a floor so the diamond dims without ever going dark.
 *
 * That floor is what separates *paused on you* from *still working*, which is
 * why it is a generated constant rather than a rounded-off literal.
 */
export function waitingBrightness(tick: number): number {
  return (
    ANIMATION.waiting_floor +
    pulseBrightness(tick, ANIMATION.waiting_pulse_speed) * ANIMATION.waiting_range
  );
}

/**
 * One shared tick, read as a signal.
 *
 * Derived from wall-clock rather than counted per frame, so a page that is
 * throttled in a background tab resumes in phase with the terminal instead of
 * lagging by however many frames it missed.
 */
export function tickAt(nowMs: number): number {
  return (nowMs / 1000) * TICKS_PER_SECOND;
}

/**
 * A shared tick signal at the pager's cadence, started lazily and stopped when
 * nothing reads it.
 *
 * One `requestAnimationFrame` loop for the whole page: everything animated in
 * the pager shares a single tick counter, and sharing it here is what keeps a
 * row of bullets pulsing in unison rather than each on its own phase.
 */
export function createTick(): () => number {
  let frame = 0;
  const [tick, setTick] = createSignal(tickAt(performance.now()));
  const loop = (): void => {
    setTick(tickAt(performance.now()));
    frame = requestAnimationFrame(loop);
  };
  onMount(() => {
    frame = requestAnimationFrame(loop);
  });
  onCleanup(() => cancelAnimationFrame(frame));
  return tick;
}

/**
 * Blend from a background toward a colour, the way the pager's `blend_color`
 * does: a per-channel lerp where `0` is the background and `1` is the colour.
 *
 * Expressed as `color-mix` so the inputs stay CSS custom properties and no
 * channel arithmetic happens here — the palette is still the only source of
 * the two endpoints.
 */
export function blendToward(background: string, color: string, brightness: number): string {
  const pct = Math.round(Math.max(0, Math.min(1, brightness)) * 100);
  return `color-mix(in srgb, ${color} ${pct}%, ${background})`;
}

File: crates/codegen/xai-grok-pager-render/src/theme/animation.rs
//! Turn animation: the curves the pager animates over, and the numbers driving them.
//!
//! ## Why this sits beside the palette
//!
//! The palette answers *what colour*; this answers *how bright, this frame*.
//! Both are the pager's alone, and a second client that derives one and
//! hand-copies the other has only moved the drift somewhere less visible. So
//! the constants live here, next to the curves that read them, and
//! [`super::tokens`] serialises both into one generated artifact under one
//! guard.
//!
//! ## What stays a user knob
//!
//! `AnimationConfig::fps` and `AnimationConfig::wave_rows` are read from
//! `pager.toml`, not from here. The first is the frame cadence, the second is
//! how far the wave's phase travels per *terminal row* — between them, exactly
//! what `settings::registry::SettingSurface::Terminal` names as terminal
//! front-end tuning ("a column width, the frame cadence"). Only the shipped
//! cadence is exported, and only so a browser has a tick to count in;
//! `wave_rows` is not, because a browser has no row grid for it to mean
//! anything in.

use documented::DocumentedFields;

/// Every constant the pager's turn animation runs on.
///
/// A struct rather than loose `const`s so [`super::tokens`] can destructure it
/// **without a `..` rest pattern**: a knob added here does not compile until it
/// has a name on the wire. That is the same load-bearing half of the drift
/// guard `tokens::color_roles` carries for colour roles — the test catches a
/// stale artifact, the pattern catches a number the browser never heard of.
#[derive(Debug, Clone, Copy, PartialEq, DocumentedFields)]
pub struct AnimationConstants {
    /// Radians per tick for the running accent wave: tool rails, running
    /// bullets, and the running verb-group diamond.
    ///
    /// A rail travels one full wave every `2π / wave_speed` ticks.
    pub wave_speed: f32,
    /// Radians per tick for every "waiting on you" diamond, whatever is being
    /// waited on.
    ///
    /// The pulse is `sin²`, whose period is π, so at 30 ticks per second this
    /// is about a 1.3s cycle.
    pub waiting_pulse_speed: f32,
    /// Brightness the waiting pulse never drops below.
    ///
    /// The diamond dims at the trough instead of going dark, which is what
    /// keeps "paused on you" legible.
    pub waiting_floor: f32,
    /// Brightness the waiting pulse adds above the floor at its peak; floor
    /// plus range is full accent.
    pub waiting_range: f32,
}

/// The pager's turn animation, as shipped.
pub const ANIMATION: AnimationConstants = AnimationConstants {
    wave_speed: 0.15,
    waiting_pulse_speed: 0.08,
    waiting_floor: 0.3,
    waiting_range: 0.7,
};

/// Compute animated brightness for a wave traveling along the accent line.
///
/// Each row has a fixed phase offset (`wave_rows` rows per full cycle), so the wave moves smoothly regardless of block height.
/// `tick` is the frame counter, `speed` is radians per tick (e.g. 0.15); returns brightness in [0.0, 1.0].
pub fn wave_brightness(tick: u64, row: u16, wave_rows: u16, speed: f32) -> f32 {
    use std::f32::consts::PI;

    let rows_per_wave = wave_rows.max(1) as f32;
    let phase = (row as f32 / rows_per_wave) * 2.0 * PI;

    let t = tick as f32 * speed;

    // sin²(t + phase) gives smooth 0-1 oscillation
    let sin_val = (t + phase).sin();
    sin_val * sin_val
}

/// Compute a pulsing brightness in [0.0, 1.0] for a single element (icon, indicator); everything sharing the same tick pulses in unison.
///
/// `tick` is the frame counter, `speed` is radians per tick, and `sin²` has period π, so one full pulse takes `π / (speed * fps)` seconds.
/// At 30fps, `speed = 0.08` gives about a 1.3s cycle; for a 2.5s cycle pass about `0.042`.
pub fn pulse_brightness(tick: u64, speed: f32) -> f32 {
    let t = tick as f32 * speed;
    let sin_val = t.sin();
    sin_val * sin_val
}

/// Brightness of a "waiting on you" cue: [`pulse_brightness`] lifted onto
/// [`AnimationConstants::waiting_floor`].
///
/// The only place the floor and the range are applied, so the pager's diamonds
/// and a browser's cannot end up with different troughs.
pub fn waiting_brightness(tick: u64) -> f32 {
    ANIMATION.waiting_floor
        + pulse_brightness(tick, ANIMATION.waiting_pulse_speed) * ANIMATION.waiting_range
}
